Design/indexing_rollout_notes.md

```diff
@@ -0,0 +1,209 @@
+The only argument for keeping it is consistency.
+Throughout this document the better term *spectrum index* or simply *index* will be used.
+The nomenclature in the Mantid code base could be changed together with this major change.
+Note that this does not necessarily imply that the nomenclature on user-interface side changes as well.
+
+## Statistics
+
+Most relevant methods that need to be removed are part of `ISpectrum`.
+The following table gives and estimate of the number of occurrences of the respective methods.
+
+| Method        | Occurrences | Files  |
+| ------------- |:-------------:|:-----:|
+| `ISpectrum::getSpectrumNo()` | 271 | 99 |
+| `ISpectrum::setSpectrumNo()` | 153 | 104 |
+| `ISpectrum::copyInfoFrom()` | 26 | 17 |
+| `ISpectrum::ISpectrum()` | ? | ? |
+| `ISpectrum::operator=()` | ? | ? |
+| `???DetectorID?()` | ~900 | ~300 |
+| `EventList::operator+=(const EventList &)` | ? | ? |
+
+An initial survey shows that almost half of the occurrences are for workspace initialization.
+Those cases are usually comparatively easy and straightforward to replace by the new style of workspace factory functions defined in `DataObjects/WorkspaceCreation.h` that use `IndexInfo` to create a workspace.
+
+Other significant fractions of the occurrences are made up by logging and exceptions and by tests that initialize individual spectra.
+Furthermore there are comparisons between the respective spectrum and detector information in two workspaces.
+
+
+## General goals
+
+- Quite a few algorithms take an input workspace, extract information about spectrum numbers and detector IDs, and initialize a new workspace based on those data. Example include `SumSpectra`, `DiffractionFocussing`, and `ExtractSpectra`.
+  One goal of the indexing library is to remove explicit uses of spectrum numbers and detector IDs in as much code as possible.
+  Since we need to refactor algorithms similar to those mentioned above, it is probably a good idea to do it properly right away.
+  That is, we should *not* just refactor the manipulation of spectrum numbers and detector IDs via `IndexInfo`.
+  Instead, we should provide functionality for grouping and extraction of spectra to create a new `IndexInfo`, without the need for explicit handling of spectrum numbers and detector IDs.
+
+  As an example, consider
+
+  ```cpp
+  namespace Indexing {
+  IndexInfo extract(const IndexInfo &, const SpectrumIndexSet &);
+  IndexInfo group(const IndexInfo &, const std::vector<SpectrumIndexSet> &);
+  }
+  ```
+
+- Should we set default detector IDs?
+  This is was workspaces are doing currently, but it does not really make sense.
+  Probably we should find a better way if setting sensible detector IDs once an instrument has been set.
+  This is something that potentially needs to be decided together with other decisions regarding Instrument-2.0.
+
+
+## NearestNeighbours
+
+- Uses `specnum_t` as map key.
+- Should simply use spectrum index?
```

Comment:
Well, it depends. If the spectrum index is only used as a map index to obtain all neighbors of *local indices* (the neighbors themselves may be non-local) it would be fine.